pkg/gofr/datasource/file/sftp/file.go

```diff
@@ -0,0 +1,142 @@
+
+type file struct {
+	*sftp.File
+	logger Logger
+}
+
+type textReader struct {
+	scanner *bufio.Scanner
+	logger  Logger
+}
+
+type jsonReader struct {
+	decoder *json.Decoder
+	token   json.Token
+}
+
+// ReadAll reads either json, csv or text fileSystem, file with multiple rows, objects or single object can be read
+// in the same way.
+// File format is decided based on the extension
+// JSON fileSystem are read in struct, while CSV fileSystem are read in pointer to string.
+//
+// newCsvFile, _ = fileStore.Open("file.csv")
+// reader := newCsvFile.ReadAll()
+//
+// Reading JSON fileSystem
+//
+//	for reader.Next() {
+//		var u User
+//		reader.Scan(&u)
+//	}
+//
+// Reading CSV fileSystem
+//
+//	for reader.Next() {
+//		    var content string
+//		    reader.Scan(&u)
+//	}
+func (f file) ReadAll() (File.RowReader, error) {
+	if strings.HasSuffix(f.File.Name(), ".json") {
+		return f.createJSONReader()
+	}
+
+	return f.createTextCSVReader(), nil
+}
+
+// Factory method to create the appropriate JSON reader.
+func (f file) createJSONReader() (File.RowReader, error) {
+	decoder := json.NewDecoder(f.File)
+
+	token, err := f.peekJSONToken(decoder)
+	if err != nil {
+		f.logger.Errorf("failed to decode JSON token %v", err)
+		return nil, err
+	}
+
+	if d, ok := token.(json.Delim); ok && d == '[' {
+		// JSON array
+		return &jsonReader{decoder: decoder, token: token}, nil
+	}
+
+	// JSON object
+	return f.createJSONObjectReader()
+}
+
+// Peek the first JSON token to determine its type.
+func (file) peekJSONToken(decoder *json.Decoder) (json.Token, error) {
+	newDecoder := *decoder
+
+	token, err := newDecoder.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+// Create a JSON reader for a JSON object.
+func (f file) createJSONObjectReader() (File.RowReader, error) {
+	name := f.File.Name()
+
+	if err := f.File.Close(); err != nil {
+		f.logger.Errorf("failed to close JSON file for reading as object %v", err)
+		return nil, err
+	}
+
+	newFile, err := os.Open(name)
+	if err != nil {
+		f.logger.Errorf("failed to open JSON file for reading as object %v", err)
+		return nil, err
+	}
+
+	decoder := json.NewDecoder(newFile)
+
+	return &jsonReader{decoder: decoder}, nil
+}
+
+func (f file) createTextCSVReader() File.RowReader {
+	return &textReader{
+		scanner: bufio.NewScanner(f.File),
+		logger:  f.logger,
+	}
+}
+
+// Next checks if there is next json object available otherwise returns false.
+func (j jsonReader) Next() bool {
+	return j.decoder.More()
+}
+
+// Scan binds the data to provided struct.
+func (j jsonReader) Scan(i interface{}) error {
+	return j.decoder.Decode(&i)
+}
+
+// Next checks if there is data available in next line otherwise returns false.
+func (f textReader) Next() bool {
+	return f.scanner.Scan()
+}
+
+// Scan binds the line to provided pointer to string.
+func (f textReader) Scan(i interface{}) error {
+	switch target := i.(type) {
+	case *string:
+		*target = f.scanner.Text()
```

Comment:
target is a local variable, with a string type.

I'm unsure what the code is supposed to do.

I mean storing something in the pointer of a variable that will be no longer exist when this code will return nil

I'm insure this would work, but I would have understand if the code had been something like that

```suggestion
		*i = f.scanner.Text()
```

Reply:
This would work because - we are checking if the `i` variable is a pointer and now target would store the same location so when we update the content at that memory it is also reflected by `i`. 

i would anyway be a pointer, Keeping it this way, it is made sure that we are assigning to a pointer